sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java

```diff
@@ -61,126 +75,263 @@ public final class AccessTokenUtil {
             result = getAccessTokenOnAppService(resource, identity);
         } else {
             result = getAccessTokenOnOthers(resource, identity);
         }
+
         return result;
     }
 
     /**
      * Get an access token.
      *
-     * @param resource             the resource.
-     * @param tenantId             the tenant ID.
-     * @param aadAuthenticationUrl the AAD authentication url
-     * @param clientId             the client ID.
-     * @param clientSecret         the client secret.
-     * @return the authorization token.
-     */
-    public static AccessToken getAccessToken(String resource, String aadAuthenticationUrl,
-               String tenantId, String clientId, String clientSecret) {
-
-        LOGGER.entering("AccessTokenUtil", "getAccessToken", new Object[]{
-            resource, tenantId, clientId, clientSecret});
+     * @param resource The resource.
+     * @param tenantId The tenant ID.
+     * @param aadAuthenticationUrl The AAD authentication url.
+     * @param clientId The client ID.
+     * @param clientSecret The client secret.
+     *
+     * @return The authorization token.
+     */
+    public static AccessToken getAccessToken(String resource, String aadAuthenticationUrl, String tenantId,
+        String clientId, String clientSecret) {
+
+        LOGGER.entering("AccessTokenUtil", "getAccessToken",
+            new Object[] { resource, tenantId, clientId, clientSecret });
         LOGGER.info("Getting access token using client ID / client secret");
+
         AccessToken result = null;
 
         StringBuilder oauth2Url = new StringBuilder();
-        oauth2Url.append(aadAuthenticationUrl == null ? OAUTH2_TOKEN_BASE_URL : aadAuthenticationUrl)
-                 .append(tenantId)
-                 .append(OAUTH2_TOKEN_POSTFIX);
 
-        StringBuilder requestBody = new StringBuilder();
-        try {
-            requestBody.append(GRANT_TYPE_FRAGMENT)
-                .append(CLIENT_ID_FRAGMENT).append(clientId)
-                .append(CLIENT_SECRET_FRAGMENT).append(URLEncoder.encode(clientSecret, "UTF-8"))
-                .append(RESOURCE_FRAGMENT).append(resource);
-        } catch (UnsupportedEncodingException e) {
-            LOGGER.warning("Failed to construct requestBody");
+        if (aadAuthenticationUrl == null) {
+            oauth2Url.append(OAUTH2_TOKEN_BASE_URL).append(tenantId).append("/");
+        } else {
+            oauth2Url.append(addTrailingSlashIfRequired(aadAuthenticationUrl));
         }
 
-        String body = HttpUtil
-            .post(oauth2Url.toString(), requestBody.toString(), "application/x-www-form-urlencoded");
+        oauth2Url.append(OAUTH2_TOKEN_POSTFIX);
+
+        StringBuilder requestBody = new StringBuilder();
+
+        requestBody.append(GRANT_TYPE_FRAGMENT)
+            .append(CLIENT_ID_FRAGMENT).append(clientId)
+            .append(CLIENT_SECRET_FRAGMENT).append(clientSecret)
```

Comment:
Seems like the encoding was added by https://github.com/Azure/azure-sdk-for-java/pull/40697. cc @rujche

Reply:
I added the encoding of the client secret, but changed the logic about handling an error a bit to avoid sending a request missing the resource.

Reply:
I saw Rujun's [comment](https://github.com/Azure/azure-sdk-for-java/pull/40697#discussion_r1645339548) about encoding every single parameter and using `scope` instead of `resource`. Is there a particular reason we didn't follow that route after all and just encode the client secret?

Reply:
1. Using `resource` instead of `scope`. I see the [doc ](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#first-case-access-token-request-with-a-shared-secret) shows the parameter name is `scope`, but original code used `resource`. I don't know why it used `resource` in original code. In my PR, I just keep original behavior: use `resource`.
2. Encoding only `clientSecret` instead of all parameters. The doc shows `The client secret must be URL-encoded before being sent`, it didn't say whether other parameters should be URL-encoded or not. I tried encode all parameters, it failed to get access token. Here is related screenshot about the doc:
> ![image](https://github.com/Azure/azure-sdk-for-java/assets/171773178/01ef48b0-1e2c-44d2-8a88-252cc8cc4c65)

Reply:
I see, we'll keep things as they are then.